File: src/Automator.Web/Services/DependencyCheckService.cs
using System.ComponentModel;
using System.Diagnostics;
using System.Runtime.InteropServices;
using Automator.Web.Data;
using Automator.Web.Models;
using Microsoft.EntityFrameworkCore;

namespace Automator.Web.Services;

public class DependencyCheckService : IDependencyCheckService
{
    private readonly IDbContextFactory<AutomatorDbContext> _dbFactory;
    private readonly IConfiguration _config;

    private static readonly (string Name, string Description, string Command, string VersionArgs, ScriptLanguage Language)[] RuntimeChecks =
    [
        ("Bash",            "Bourne Again Shell",               "bash",             "--version",  ScriptLanguage.Bash),
        ("PowerShell Core", "Cross-platform PowerShell (pwsh)", "pwsh",             "--version",  ScriptLanguage.PowerShell),
        ("Python 3",        "Python interpreter",               "python3",          "--version",  ScriptLanguage.Python),
        ("Ansible",         "Ansible automation platform",      "ansible-playbook", "--version",  ScriptLanguage.Ansible),
        ("Terraform",       "Infrastructure as Code tool",      "terraform",        "version",    ScriptLanguage.Terraform),
    ];

    public DependencyCheckService(IDbContextFactory<AutomatorDbContext> dbFactory, IConfiguration config)
    {
        _dbFactory = dbFactory;
        _config = config;
    }

    public async Task<IReadOnlyList<DependencyCheckResult>> CheckRuntimesAsync(CancellationToken ct = default)
    {
        var tasks = RuntimeChecks.Select(r => CheckRuntimeAsync(r.Name, r.Description, r.Command, r.VersionArgs, r.Language, ct));
        return await Task.WhenAll(tasks);
    }

    public async Task<DatabaseStatus> GetDatabaseStatusAsync(CancellationToken ct = default)
    {
        var connectionString = _config.GetConnectionString("DefaultConnection") ?? "Data Source=automator.db";
        var filePath = connectionString
            .Split(';', StringSplitOptions.RemoveEmptyEntries)
            .Select(p => p.Trim())
            .FirstOrDefault(p => p.StartsWith("Data Source=", StringComparison.OrdinalIgnoreCase))
            ?.Substring("Data Source=".Length)
            ?? "automator.db";

        var fullPath = Path.IsPathRooted(filePath)
            ? filePath
            : Path.Combine(Directory.GetCurrentDirectory(), filePath);

        long? fileSize = File.Exists(fullPath) ? new FileInfo(fullPath).Length : null;

        try
        {
            await using var db = await _dbFactory.CreateDbContextAsync(ct);
            var scriptCount = await db.Scripts.CountAsync(ct);
            var jobCount = await db.ScheduledJobs.CountAsync(ct);
            var historyCount = await db.ExecutionHistory.CountAsync(ct);

            return new DatabaseStatus(fullPath, fileSize, true, scriptCount, jobCount, historyCount);
        }
        catch (Exception ex)
        {
            return new DatabaseStatus(fullPath, fileSize, false, 0, 0, 0, ex.Message);
        }
    }

    private static async Task<DependencyCheckResult> CheckRuntimeAsync(
        string name, string description, string command, string versionArgs,
        ScriptLanguage language, CancellationToken ct)
    {
        var resolvedCommand = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && language == ScriptLanguage.PowerShell
            ? "powershell.exe"
            : command;

        try
        {
            using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            timeoutCts.CancelAfter(TimeSpan.FromSeconds(5));

            var path = await ResolvePathAsync(resolvedCommand, timeoutCts.Token);

            var psi = new ProcessStartInfo
            {
                FileName = resolvedCommand,
                Arguments = versionArgs,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using var process = Process.Start(psi) ?? throw new InvalidOperationException("Process did not start");

            var stdoutTask = process.StandardOutput.ReadToEndAsync(timeoutCts.Token);
            var stderrTask = process.StandardError.ReadToEndAsync(timeoutCts.Token);
            await process.WaitForExitAsync(timeoutCts.Token);

            var output = await stdoutTask + await stderrTask;
            var version = output.Split('\n', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault()?.Trim();

            return new DependencyCheckResult(name, description, true, version, path, null, language);
        }
        catch (OperationCanceledException)
        {
            return new DependencyCheckResult(name, description, false, null, null, "Timed out after 5s", language);
        }
        catch (Exception ex) when (ex is Win32Exception or FileNotFoundException or InvalidOperationException)
        {
            return new DependencyCheckResult(name, description, false, null, null, "Not found in PATH", language);
        }
        catch (Exception ex)
        {
            return new DependencyCheckResult(name, description, false, null, null, ex.Message, language);
        }
    }

    private static async Task<string?> ResolvePathAsync(string command, CancellationToken ct)
    {
        var whichCommand = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "where" : "which";
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = whichCommand,
                Arguments = command,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            using var process = Process.Start(psi);
            if (process is null) return null;
            var output = await process.StandardOutput.ReadLineAsync(ct);
            await process.WaitForExitAsync(ct);
            return output?.Trim();
        }
        catch { return null; }
    }
}

File: automator/app/Services/DependencyCheckService.php
<?php

namespace App\Services;

use App\Enums\ScriptLanguage;
use Illuminate\Support\Facades\DB;
use Symfony\Component\Process\Exception\ProcessTimedOutException;
use Symfony\Component\Process\Process;

class DependencyCheckService
{
    /**
     * @return array{name: string, description: string, language: ?ScriptLanguage, command: string, version_args: string}[]
     */
    private const RUNTIME_CHECKS = [
        ['name' => 'Bash', 'description' => 'Bourne Again Shell', 'command' => 'bash', 'version_args' => '--version', 'language' => ScriptLanguage::Bash],
        ['name' => 'PowerShell Core', 'description' => 'Cross-platform PowerShell (pwsh)', 'command' => 'pwsh', 'version_args' => '--version', 'language' => ScriptLanguage::PowerShell],
        ['name' => 'Python 3', 'description' => 'Python interpreter', 'command' => 'python3', 'version_args' => '--version', 'language' => ScriptLanguage::Python],
        ['name' => 'Ansible', 'description' => 'Ansible automation platform', 'command' => 'ansible-playbook', 'version_args' => '--version', 'language' => ScriptLanguage::Ansible],
        ['name' => 'Terraform', 'description' => 'Infrastructure as Code tool', 'command' => 'terraform', 'version_args' => 'version', 'language' => ScriptLanguage::Terraform],
    ];

    /**
     * @return array{name: string, description: string, available: bool, version: ?string, path: ?string, error: ?string, language: ScriptLanguage}[]
     */
    public function checkRuntimes(): array
    {
        return array_map(
            fn (array $check) => $this->checkRuntime(...$check),
            self::RUNTIME_CHECKS,
        );
    }

    private function checkRuntime(string $name, string $description, string $command, string $version_args, ScriptLanguage $language): array
    {
        try {
            $path = $this->resolvePath($command);

            $process = new Process([$command, ...explode(' ', $version_args)]);
            $process->setTimeout(5);
            $process->run();

            $output = $process->getOutput().$process->getErrorOutput();
            $version = collect(preg_split('/\r?\n/', trim($output)))->first();

            return [
                'name' => $name,
                'description' => $description,
                'available' => $process->isSuccessful() || $path !== null,
                'version' => $version ?: null,
                'path' => $path,
                'error' => null,
                'language' => $language,
            ];
        } catch (ProcessTimedOutException) {
            return ['name' => $name, 'description' => $description, 'available' => false, 'version' => null, 'path' => null, 'error' => 'Timed out after 5s', 'language' => $language];
        } catch (\Throwable $e) {
            return ['name' => $name, 'description' => $description, 'available' => false, 'version' => null, 'path' => null, 'error' => 'Not found in PATH', 'language' => $language];
        }
    }

    private function resolvePath(string $command): ?string
    {
        try {
            $process = new Process(['which', $command]);
            $process->setTimeout(5);
            $process->run();

            return $process->isSuccessful() ? trim($process->getOutput()) ?: null : null;
        } catch (\Throwable) {
            return null;
        }
    }

    /**
     * @return array{driver: string, database: ?string, connected: bool, error: ?string, script_count: int, job_count: int, history_count: int}
     */
    public function databaseStatus(): array
    {
        $connection = config('database.default');
        $database = config("database.connections.{$connection}.database");

        try {
            DB::connection()->getPdo();

            return [
                'driver' => $connection,
                'database' => is_string($database) ? $database : null,
                'connected' => true,
                'error' => null,
                'script_count' => \App\Models\ScriptDefinition::count(),
                'job_count' => \App\Models\ScheduledJob::count(),
                'history_count' => \App\Models\ScriptExecutionResult::count(),
            ];
        } catch (\Throwable $e) {
            return [
                'driver' => $connection,
                'database' => is_string($database) ? $database : null,
                'connected' => false,
                'error' => $e->getMessage(),
                'script_count' => 0,
                'job_count' => 0,
                'history_count' => 0,
            ];
        }
    }

    /**
     * @return array{php_version: string, laravel_version: string, os: string, queue_driver: string, server_time: string}
     */
    public function appInfo(): array
    {
        return [
            'php_version' => PHP_VERSION,
            'laravel_version' => app()->version(),
            'os' => PHP_OS_FAMILY,
            'queue_driver' => config('queue.default'),
            'server_time' => now()->format('Y-m-d H:i:s T'),
        ];
    }
}
